cv32/bsp/vectors.S

```diff
@@ -22,32 +22,33 @@ vector_table:
 	j sw_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
+	j software_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
+	j timer_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
+	j external_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
 	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j __no_irq_handler
-	j verification_irq_handler
+	j fast0_irq_handler
+	j fast1_irq_handler
+	j fast2_irq_handler
+	j fast3_irq_handler
+	j fast4_irq_handler
+	j fast5_irq_handler
+	j fast6_irq_handler
+	j fast7_irq_handler
+	j fast8_irq_handler
+	j fast9_irq_handler
+	j fast10_irq_handler
+	j fast11_irq_handler
+	j fast12_irq_handler
+	j fast13_irq_handler
+	j fast14_irq_handler
+	j fast15_irq_handler
+
```

Comment:
My thought is that if you are going to make this change, you should do it for all handlers, not just fast IRQ handlers.

Overall, adding explicit handlers is fine. But I don't think they should be weak references (certainly not for the verification use case). And if you want to have explicit handlers, make them all explicit.

Within this patch, the question of what appears to be duplication of the software IRQ handler has to be addressed. (see earlier comment).

Reply:
I would still really like for the handlers to be weak.  Referencing above in certain test scenarios (i.e. the interrupt test) I would add items to a handler that would simply be overhead in most other contexts.  

Please let me know if you know of an alternative to weak.  Is adding an optional callback preferable in the BSP?

Reply:
I'll chime in on the handlers.  In verification we need the ability to deliberately break things and model (or react to) the expected response of the DUT when we do.  Typically this is done with callbacks or "hook" methods.  This is my first experience trying to do this with assembly, so any implementation advice is welcome, as long as we have ability to modify the "correct" behavior of the BSP so that we can abuse the DUT.

Reply:
Given there's a use case, I'm happy for them to be weak unless @jeremybennett would like to recommend an alternative.

Reply:
I'm really uncomfortable with weak references for interrupt handlers.  I prefer Mike's approach. which is less tied to the specifics of ELF linking on Linux.  Create a set of default handlers using an indirection table. Something like (in pseudo-C)
```c
static void (*ext_handlers[NUM_INTERRUPTS]) (int a1, int a2, int a3) =
{ NULL, NULL, ... };


void generic_handler (unsigned int  num, int a1, int a2, int a3)
{
  if (ext_handlers[num])
    *(ext_handlers[num]) ()
  else
    <default handler>
}

void handler_callback (void (*handler) (unsigned int num, int a1, int a2, int a3)
{
  ext_handlers[num] = handler;
}
```
So your handler will do some default action, unless (presumably at start up), you call `handler_callback ()` to add a custom handler.  This code is somewhat simplified - you'd need error checking etc, but it shows the way.